pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,645 @@
+	TopicOptions TopicOptions
+	// SubscriptionOptions specifies the options to pass to OpenSubscription.
+	SubscriptionOptions SubscriptionOptions
+}
+
+// OpenTopicURL opens a pubsub.Topic based on u.
+func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic, error) {
+	for param := range u.Query() {
+		return nil, fmt.Errorf("open topic %v: invalid query parameter %s", u, param)
+	}
+	subject := path.Join(u.Host, u.Path)
+	return OpenTopic(o.Connection, subject, &o.TopicOptions)
+}
+
+// OpenSubscriptionURL opens a pubsub.Subscription based on input url.
+func (o *URLOpener) OpenSubscriptionURL(ctx context.Context, u *url.URL) (*pubsub.Subscription, error) {
+	opts := o.SubscriptionOptions
+	for param, values := range u.Query() {
+		switch param {
+		case "queue":
+			opts.Queue = values[0]
+		case "durable_name":
+			opts.DurableName = values[0]
+		case "max_inflight":
+			maxInflight, err := strconv.Atoi(values[0])
+			if err != nil {
+				return nil, fmt.Errorf("open subscription %v: invalid value for maxInflight query parameter: %s, %w", u, values[0], err)
+			}
+			opts.MaxInflight = maxInflight
+		case "start_sequence":
+			startSequence, err := strconv.ParseUint(values[0], 10, 64)
+			if err != nil {
+				return nil, fmt.Errorf("open subscription %v: invalid value for startSequence query parameter: %s, %w", u, values[0], err)
+			}
+			opts.StartSequence = startSequence
+			opts.StartAt = StartPositionSequenceStart
+		case "ack_wait":
+			ackWait, err := time.ParseDuration(values[0])
+			if err != nil {
+				return nil, fmt.Errorf("open subscription %v: invalid value for ackWait query parameter: %s, %w", u, values[0], err)
+			}
+			opts.AckWait = ackWait
+		case "manual_acks":
+			if values[0] != "" {
+				return nil, fmt.Errorf("open subscription %v: no values are permitted for the manualAck query parameter", u)
+			}
+			opts.ManualAcks = true
+		default:
+			return nil, fmt.Errorf("open subscription %v: invalid query parameter %s", u, param)
+		}
+	}
+	subject := path.Join(u.Host, u.Path)
+
+	sub, err := openSubscription(o.Connection, subject, &opts)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewSubscription(sub, &batcher.Options{MaxBatchSize: opts.MaxInflight}, &batcher.Options{MaxBatchSize: opts.MaxInflight}), nil
+}
+
+// TopicOptions sets options for constructing a *pubsub.Topic backed by NATS Streaming.
+type TopicOptions struct{}
+
+// OpenTopic returns a *pubsub.Topic for use with NATS.
+// The subject is the NATS Subject; for more info, see
+// https://nats.io/documentation/writing_applications/subjects.
+func OpenTopic(nc stan.Conn, subject string, _ *TopicOptions) (*pubsub.Topic, error) {
+	dt, err := openTopic(nc, subject)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(dt, nil), nil
+}
+
+// SubscriptionOptions sets options for constructing a *pubsub.Subscription backed by NATS Streaming.
+type SubscriptionOptions struct {
```

Comment:
Just to be clear, all of these are optional, right?

Reply:
Yeah, exactly. All of these are optional.